### PR TITLE
Broader Runnable support, compositional pieces

### PR DESCRIPTION
**Pull request by jacoblee93**
Adds RunnableSequence, RunnableMap, RunnableLambda, and RunnablePassthrough and associated compositional `.pipe` methods, also makes retrievers and output parsers runnable.

### CODE REVIEW

File: langchain/src/schema/output_parser.ts [hunk 1]
```diff
@@ -20,9 +28,31 @@ export abstract class BaseLLMOutputParser<T = unknown> extends Serializable {
...
+    input: string | BaseMessage,
+    options?: RunnableConfig
+  ): Promise<T> {
+    if (typeof input === "string") {
+      return this._callWithConfig(
+        async (input: string): Promise<T> =>
+          this.parseResult([{ text: input }]),
+        input,
+        options
+      );
```
Comment (nfcampos) [inline langchain/src/schema/output_parser.ts:42]: Add here run type per https://github.com/langchain-ai/langchain/pull/8488

File: langchain/src/schema/runnable.ts [hunk 2]
```diff
@@ -82,4 +106,401 @@ export abstract class Runnable<
...
+      });
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  static isInstance(thing: any): thing is RunnableSequence {
+    return Array.isArray(thing.middle) && Runnable.isInstance(thing);
+  }
+
+  static fromRunnables<RunInput, RunOutput>([first, ...runnables]: [
```
Comment (nfcampos) [inline langchain/src/schema/runnable.ts:376]: Rename to `.from()` ? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/from

File: langchain/src/schema/runnable.ts [hunk 3]
```diff
@@ -82,4 +106,401 @@ export abstract class Runnable<
...
+  }
+}
+
+export class RunnableLambda<RunInput, RunOutput> extends Runnable<
+  RunInput,
+  RunOutput
+> {
+  lc_namespace = ["schema", "runnable"];
+
+  constructor(protected func: RunnableFunc<RunInput, RunOutput>) {
```
Comment (nfcampos) [inline langchain/src/schema/runnable.ts:449]: This needs to be an object with single key func

File: langchain/src/schema/runnable.ts [hunk 4]
```diff
@@ -82,4 +106,401 @@ export abstract class Runnable<
...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Record<string, any>
+> {
+  lc_namespace = ["schema", "runnable"];
+
+  lc_serializable = true;
+
+  protected steps: Record<string, Runnable<RunInput>>;
+
+  constructor(steps: Record<string, RunnableLike<RunInput>>) {
```
Comment (nfcampos) [inline langchain/src/schema/runnable.ts:401]: This needs to be an object with single key `steps`. Otherwise serialisation won't match py

Comment (jacoblee93) [reply langchain/src/schema/runnable.ts:401]: Ah right, thank you!

**Pull request merged**